File: appmy/src/main/java/com/leilei/guoshujinfu/mylearning/util/BaseFragment.java
package com.leilei.guoshujinfu.mylearning.util;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * @Author: AqrLei
 * @Name MyLearning
 * @Description:
 * @Date: 2017/8/3
 */

public abstract class BaseFragment<T extends BaseActivity> extends Fragment {
    protected  T mContainerActivity;
    protected View mView;
    private Unbinder unbinder;

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if(context instanceof BaseActivity) {
            mContainerActivity = ((T)context);
            AppLog.logDebug("context", ""+ mContainerActivity.getClass().getSimpleName());
           if (mContainerActivity == null)
            AppLog.logDebug("context", "context is null");
        }
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
                             Bundle savedInstanceState) {
        mView = inflater.inflate(getLayoutRes(), container, false);
        unbinder = ButterKnife.bind(this, mView);
        initComponents();
        return mView;


    }

    @Override
    public void onDetach() {
        super.onDetach();
        if(mContainerActivity != null) {
            mContainerActivity = null;
        }
    }
    public T getContainerActivity() {
        return mContainerActivity;
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    protected abstract int getLayoutRes();
    protected void initComponents() {

    }
    public void finish () {
        mContainerActivity.finish();
    }



}

File: aqrleilib/src/main/kotlin/com/aqrairsigns/aqrleilib/basemvp/BaseFragment.kt
package com.aqrairsigns.aqrleilib.basemvp

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.aqrairsigns.aqrleilib.util.AppLog


/**
 * @Author: AqrLei
 *@Name MyLearning
 *@Description:
 *@Date: 2017/8/22
 */
abstract class BaseFragment<T : BaseActivity> : Fragment() {
    protected lateinit var mContainerActivity: T
    protected lateinit var mView: View


    protected abstract val layoutRes: Int
    protected open fun initComponents(view: View?, savedInstanceState: Bundle?) {

    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is BaseActivity) {
            @Suppress("UNCHECKED_CAST")
            mContainerActivity = context as T
            AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onAttach----" +
                    mContainerActivity.javaClass.simpleName)

        }
    }

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        mView = inflater!!.inflate(layoutRes, container, false)
        //initComponents()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onCreateView----" +
                mContainerActivity.javaClass.simpleName)
        return mView
    }

    override fun onViewCreated(view: View?, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initComponents(view, savedInstanceState)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onActivityCreated----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onStart() {
        super.onStart()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onStart----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onResume() {
        super.onResume()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onResume----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onPause() {
        super.onPause()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onPause----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onStop() {
        super.onStop()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onStop----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onDestroyView----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onDestroy() {
        super.onDestroy()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onDestroy----" +
                mContainerActivity.javaClass.simpleName)
    }

    override fun onDetach() {
        super.onDetach()
        AppLog.logDebug(AppLog.LOG_TAG_FRAGMENT, this.javaClass.simpleName + "----onDetach----" +
                mContainerActivity.javaClass.simpleName)
    }

    fun finish() {
        mContainerActivity.finish()
    }


}